### PR TITLE
Add support for renderer backoff, don't FAIL_FAST on 3x failures, add UI

### LINKED ISSUE #5340: Crash in render engine due to directx device hanging up
VSExtensibility team got a watson logged internally with a crash coming from the terminal. After looking into the watson it appears that the crash is coming from this line:

https://github.com/microsoft/terminal/blob/ca33d895a359664c337c6482b853489823c216d3/src/renderer/base/renderer.cpp#L81

It looks like an intentional fail fast but it would be better to surface the error and allow the application hosting the rendering engine to decide if it should crash or not.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
## Summary of the Pull Request
Renderer: Add support for backoff and auto-disable on failed retry

This commit introduces a backoff (150ms * number of tries) to the
renderer's retry logic (introduced in #2830). It also changes the
FAIL_FAST to a less globally-harmful render thread disable, so that we
stop blowing up any application hosting a terminal when the graphics
driver goes away.

In addition, it adds a callback that a Renderer consumer can use to
determine when the renderer _has_ failed, and a public method to kick it
back into life.

Fixes #5340.

This PR also wires up TermControl so that it shows some UI when the renderer tastes clay.

image

image

## PR Checklist
* [x] Closes #5340
* [x] cla
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already.

## Validation Steps Performed

I tested this by dropping the number of retries to 1 and forcing a TDR while doing `wsl cmatrix -u0`. It picked up exactly where it left off.

As a bonus, you can actually still type into the terminal when it's graphically suspended (and `exit` still works.). The block is _entirely graphical_.

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.xaml [hunk 1]
```diff
@@ -77,6 +77,23 @@
                                CurrentCursorPosition="_CurrentCursorPositionHandler"
                                CurrentFontInfo="_FontInfoHandler" />
 
+        <Grid x:Name="RendererFailedNotice"
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.xaml:80]: thank god you didn't use a Dialog

File: src/renderer/base/renderer.cpp [hunk 2]
```diff
@@ -78,8 +78,17 @@ Renderer::~Renderer()
...
+                    _pThread->DisablePainting();
+                    if (_pfnRendererEnteredErrorState)
+                    {
+                        _pfnRendererEnteredErrorState();
+                    }
+                    return S_FALSE;
                 }
+                // Add a bit of backoff.
+                // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
+                Sleep(150 * (maxRetriesForRenderEngine - tries));
```
Comment (zadjii-msft) [inline src/renderer/base/renderer.cpp:91]: Any particular reason for the backoff like this? just to give it another chance?

Comment (zadjii-msft) [inline src/renderer/base/renderer.cpp:91]: Should 150 be a constant somewhere? 🤷‍♂️

File: src/renderer/base/renderer.cpp [hunk 3]
```diff
@@ -78,8 +78,17 @@ Renderer::~Renderer()
...
                 if (--tries == 0)
                 {
-                    FAIL_FAST_HR_MSG(E_UNEXPECTED, "A rendering engine required too many retries.");
+                    // Stop trying.
+                    _pThread->DisablePainting();
+                    if (_pfnRendererEnteredErrorState)
+                    {
+                        _pfnRendererEnteredErrorState();
+                    }
+                    return S_FALSE;
```
Comment (zadjii-msft) [inline src/renderer/base/renderer.cpp:87]: If there's not a callback, and we `S_FALSE` here, there's effectively no way for the hosting app to recover, right? they won't even _know_ that something went wrong, right?

Should we instead still fail-fast if there's no callback?

**Comment by zadjii-msft** [comment]
@msftbot make sure @miniksa signs off on this

File: src/cascadia/TerminalControl/TermControl.xaml (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.xaml:80]: ha!

File: src/renderer/base/renderer.cpp (hunk 2 above)
Comment (DHowett-MSFT) [reply src/renderer/base/renderer.cpp:91]: Under the idea that the graphics driver probably went down for a real reason, and that restarting the card might take some time if that's what it needs to do.

File: src/renderer/base/renderer.cpp (hunk 3 above)
Comment (DHowett-MSFT) [reply src/renderer/base/renderer.cpp:87]: This is correct. The WPF control won't hook it up, for now. If we failfast here, without a callback, they'll be in the same boat as before (a failfast.)

Zoey suggested that us blacking out is fine, and the user can close/reopen the terminal if this happens.  It should be rarer with the backoff time.

**Comment by miniksa** [comment]
Looks alright, but I'd like to see @zadjii-msft's questions answered and @cinnamon-msft give this a pass for how it looks before we commit it.

File: src/renderer/base/renderer.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/renderer/base/renderer.cpp:87]: Okay if Zoey said this is okay then I'm fine with this. I figured leaving the failfast would encourage other apps (WPF) to implement a similar retry button, but if this is what they want the :shipit:

Comment (DHowett-MSFT) [reply src/renderer/base/renderer.cpp:87]: Commenting that.

**Comment by DHowett-MSFT** [comment]
@msftbot merge this in 2 minutes

**Pull request merged**